game/src/devtools/polygon.rs

```diff
@@ -39,34 +41,104 @@ impl PolygonEditor {
                 ctx.style()
                     .btn_outline
                     .text("export as an Osmosis polygon filter")
+                    // TODO Disable based on number of points
                     .hotkey(Key::X)
                     .build_def(ctx),
             ]))
             .aligned(HorizontalAlignment::Center, VerticalAlignment::Top)
             .build(ctx),
             name,
             points,
-            mouseover_pt: None,
-            moving_pt: false,
+            world: Cached::new(),
         })
     }
+
+    fn rebuild_world(&mut self, ctx: &mut EventCtx, app: &App) {
+        let mut world = World::bounded(app.primary.map.get_bounds());
+
+        if self.points.len() >= 3 {
+            let mut pts = self.points.to_vec();
+            pts.push(pts[0]);
+            world
+                .add(Obj::Polygon)
+                .hitbox(Ring::must_new(pts).into_polygon())
+                .zorder(0)
+                .draw_color(Color::BLUE.alpha(0.6))
+                .hover_alpha(0.3)
+                .draggable()
+                .build(ctx);
+        }
+
+        for (idx, pt) in self.points.iter().enumerate() {
+            world
+                .add(Obj::Point(idx))
+                // Scale the circle as we zoom out
+                .hitbox(Circle::new(*pt, Distance::meters(10.0) / ctx.canvas.cam_zoom).to_polygon())
+                .zorder(1)
+                .draw_color(Color::RED)
+                .hover_alpha(0.8)
+                .hotkey(Key::Backspace, "delete")
+                .draggable()
+                .build(ctx);
+        }
+
+        world.initialize_hover(ctx);
+
+        if let Some(prev) = self.world.value() {
+            world.rebuilt_during_drag(prev);
+        }
+        self.world.set(ctx.canvas.cam_zoom, world);
+    }
 }
 
 impl State<App> for PolygonEditor {
```

Comment:
I'm particularly happy with how this code turned out. We're almost object-oriented (in the good way). We have a simple model (list of points), a function to transform that into a view and express interactions, and all of the handlers for those high-level interactions here. In some languages/frameworks, these would be callbacks included directly in the object builder API. We don't need to go that far, but we potentially could.